lib/spack/llnl/util/filesystem.py

```diff
@@ -956,10 +956,45 @@ class HeaderList(FileList):
     commonly used compiler flags or names.
     """
 
+    include_regex = re.compile(r'(.*)(include)(.*)')
+
     def __init__(self, files):
         super(HeaderList, self).__init__(files)
 
         self._macro_definitions = []
+        self._directories = None
+
+    @property
+    def directories(self):
+        """Directories to be searched for header files."""
+        values = self._directories or self._default_directories()
+        return list(dedupe(values))
+
+    @directories.setter
+    def directories(self, value):
+        # Accept a single directory as input
+        if isinstance(value, six.string_types):
+            value = [value]
+
+        # Setting the property to None makes the initial default
+        # kick-in again
+        self._directories = None if value is None else [
```

Comment:
Good catch. The bug is above, I'd like to leave open the possibility to set an empty list so that no directory will be returned. I'll fix this and add a test.

Reply:
IMO differing behavior for `None` and `[]` is unintuitive. I think it's fine if `_directories` is assigned to `None` in `__init__`, but that the setter should either forbid assigning `None` or should handle `None` like `[]`. So an assignment like:

`self._directories = value or []`

Internally, we can use `None` to indicate that the default should be used, but once the user assigns, I think it makes sense to use that to completely override the default.